lib/editor/components/timetable/TimetableGrid.js

```diff
@@ -378,27 +378,35 @@ export default class TimetableGrid extends Component<Props> {
     } = this.props
     let activeRow = rowIndex
     let activeCol = colIndex
-    // iterate over rows in pasted selection
+    // Iterate over rows in pasted selection
     for (var i = 0; i < pastedRows.length; i++) {
       activeRow = rowIndex + i
-      // construct new row if it doesn't exist
-      if (typeof data[i + rowIndex] === 'undefined') {
+      // Construct new row if it doesn't exist
+      if (typeof data[activeRow] === 'undefined') {
         addNewRow()
       }
-      // iterate over number of columns in pasted selection
-      for (var j = 0; j < pastedRows[0].length; j++) {
+      // Iterate over number of columns in pasted selection
+      for (var j = 0; j < pastedRows[i].length; j++) {
         activeCol = colIndex + j
-        const path = `${rowIndex + i}.${columns[colIndex + j].key}`
-        const value = parseTime(pastedRows[i][j])
-        updateCellValue({value, rowIndex: rowIndex + i, key: path})
-        // if departure times are hidden, paste into adjacent time column
-        const adjacentPath = `${rowIndex + i}.${columns[colIndex + j + 2].key}`
-        if (
-          hideDepartureTimes &&
-          isTimeFormat(columns[colIndex + j].type) &&
-          typeof objectPath.get(data, adjacentPath) !== 'undefined'
-        ) {
-          updateCellValue({value, rowIndex: rowIndex + i, key: adjacentPath})
+        const col = columns[activeCol]
+        const path = `${activeRow}.${col.key}`
+        const originalValue = objectPath.get(data[activeRow], col.key)
+        const value = parseCellValue(pastedRows[i][j], col)
+        if (value !== null) {
+          const finalValue = value
+          updateCellValue({ value: finalValue, rowIndex: activeRow, key: path })
+          // If departure times are hidden, paste into adjacent time column
+          const adjacentPath = `${activeRow}.${columns[activeCol + 2].key}`
+          if (
+            hideDepartureTimes &&
+            isTimeFormat(col.type) &&
+            typeof objectPath.get(data, adjacentPath) !== 'undefined'
+          ) {
+            updateCellValue({ value: finalValue, rowIndex: activeRow, key: adjacentPath })
+          }
+        } else {
+          // If the value is rejected, keep the original value
+          updateCellValue({ value: originalValue, rowIndex: activeRow, key: path })
```

Comment:
You might need adjacentPath in the key to keep the originalValue?

Reply:
I am not experiencing the problem that the original value isn't kept yet -- for me this logic works okay